From bdf245c31bceb56a9bd8cfebca50a2b1fdb03c5c Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Thu, 6 Sep 2018 17:57:49 +0900
Subject: [PATCH 326/909] drm: rcar-du: Fix dpad signal routing for R8A7799X

commit 99547173de6e8e1ab5ef1505bf1ec0eb736c4bca from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

In the case of R8A7799X, the dpad0_source has an initial value of 0
when starting kernel, DPAD can not be set with DU1 when using with
LVDS0 -> HDMI (DU0). So correct it so that it searches and decides
connector information of VGA.

In addition, The default value of dpad0_source is determined to be 1,
and the default setting is also changed.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_crtc.c | 24 ++++++++++++++++++++++--
 drivers/gpu/drm/rcar-du/rcar_du_drv.c  |  3 ++-
 drivers/gpu/drm/rcar-du/rcar_du_drv.h  |  1 +
 drivers/gpu/drm/rcar-du/rcar_du_kms.c  |  7 ++++++-
 4 files changed, 31 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
index 7cb8cfeaa1a7..e635fc8a5d7b 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
@@ -315,6 +315,10 @@ void rcar_du_crtc_route_output(struct drm_crtc *crtc,
 {
 	struct rcar_du_crtc *rcrtc = to_rcar_crtc(crtc);
 	struct rcar_du_device *rcdu = rcrtc->group->dev;
+	struct drm_device *dev = rcrtc->crtc.dev;
+	struct drm_connector *connector;
+	struct rcar_du_crtc *tmp_rcrtc = NULL;
+	bool dpad_set = false;
 
 	/*
 	 * Store the route from the CRTC output to the DU output. The DU will be
@@ -324,9 +328,25 @@ void rcar_du_crtc_route_output(struct drm_crtc *crtc,
 
 	/*
 	 * Store RGB routing to DPAD0, the hardware will be configured when
-	 * starting the CRTC.
+	 * starting the CRTC. In the case of R8A7799X, because it is LVDS
+	 * output, locate the VGA connector and identify the CRTC.
 	 */
-	if (output == RCAR_DU_OUTPUT_DPAD0)
+	list_for_each_entry(connector,
+			    &dev->mode_config.connector_list, head) {
+		if (rcar_du_has(rcdu, RCAR_DU_FEATURE_R8A7799X) && connector &&
+		    (connector->connector_type == DRM_MODE_CONNECTOR_VGA) &&
+		    (connector->encoder)) {
+			tmp_rcrtc = to_rcar_crtc(connector->encoder->crtc);
+			break;
+		}
+	}
+
+	if (tmp_rcrtc && (rcrtc->index == tmp_rcrtc->index))
+		dpad_set = true;
+
+	if (rcar_du_has(rcdu, RCAR_DU_FEATURE_R8A7799X) && dpad_set)
+		rcdu->dpad0_source = rcrtc->index;
+	else if (output == RCAR_DU_OUTPUT_DPAD0)
 		rcdu->dpad0_source = rcrtc->index;
 }
 
diff --git a/drivers/gpu/drm/rcar-du/rcar_du_drv.c b/drivers/gpu/drm/rcar-du/rcar_du_drv.c
index 465e1f417e3f..fabb37fe0b3e 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_drv.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_drv.c
@@ -317,7 +317,8 @@ static const struct rcar_du_device_info rcar_du_r8a7799x_info = {
 	.gen = 3,
 	.features = RCAR_DU_FEATURE_CRTC_IRQ_CLOCK
 		  | RCAR_DU_FEATURE_EXT_CTRL_REGS
-		  | RCAR_DU_FEATURE_VSP1_SOURCE,
+		  | RCAR_DU_FEATURE_VSP1_SOURCE
+		  | RCAR_DU_FEATURE_R8A7799X,
 	.channels_mask = BIT(1) | BIT(0),
 	.routes = {
 		/*
diff --git a/drivers/gpu/drm/rcar-du/rcar_du_drv.h b/drivers/gpu/drm/rcar-du/rcar_du_drv.h
index 62f319c19ffd..504a188b91f7 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_drv.h
+++ b/drivers/gpu/drm/rcar-du/rcar_du_drv.h
@@ -36,6 +36,7 @@ struct rcar_du_device;
 #define RCAR_DU_FEATURE_R8A77990_REGS	(1 << 6)        /* Use R8A77990 registers */
 #define RCAR_DU_FEATURE_R8A77995_REGS	(1 << 7)        /* Use R8A77995 registers */
 #define RCAR_DU_FEATURE_TVM_SYNC	(1 << 8)	/* Has TV switch/sync modes */
+#define RCAR_DU_FEATURE_R8A7799X	(1 << 9)        /* Use R8A7799X */
 
 #define RCAR_DU_QUIRK_ALIGN_128B	(1 << 0)	/* Align pitches to 128 bytes */
 
diff --git a/drivers/gpu/drm/rcar-du/rcar_du_kms.c b/drivers/gpu/drm/rcar-du/rcar_du_kms.c
index 01c0e831820d..4cf00eee33fc 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_kms.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_kms.c
@@ -741,7 +741,12 @@ int rcar_du_modeset_init(struct rcar_du_device *rcdu)
 	dpad0_sources = rcdu->info->routes[RCAR_DU_OUTPUT_DPAD0].possible_crtcs;
 	if (rcar_du_has(rcdu, RCAR_DU_FEATURE_R8A77965_REGS))
 		dpad0_sources = dpad0_sources << 1;
-	rcdu->dpad0_source = ffs(dpad0_sources) - 1;
+
+	/* In the case of R8A7799X, default is set to 1.*/
+	if (rcar_du_has(rcdu, RCAR_DU_FEATURE_R8A7799X))
+		rcdu->dpad0_source = 1;
+	else
+		rcdu->dpad0_source = ffs(dpad0_sources) - 1;
 
 	/* Create the CRTCs. */
 	for (swindex = 0, hwindex = 0; swindex < rcdu->num_crtcs; ++hwindex) {
-- 
2.17.1

